R/modify-recursive.R

```diff
@@ -0,0 +1,68 @@
+#' Recursively modify a list
+#'
+#' `rmodify()` allows you to recursively modify a list, supplying functions
+#' that either modify each leaf or each node (or both).
+#'
+#' @param x A list.
+#' @param f_leaf A function applied to each leaf.
+#' @param p_leaf A predicate function that returns `TRUE` when an element is
+#'   a leaf, determining whether `f_leaf` or `f_pre`/`f_post` is applied to
+#'   it. The default value, `NULL`, treats lists as nodes and everything else
+#'   as leaves.
+#' @param f_pre,f_post Functions applied to each node. `f_pre` is applied
+#'   the tree is traversed "down", i.e. before the leaves are transformed
+#'   with `f_leaf`, while `f_post` is applied on the way "up", i.e.
+#'   after the leaves are transformed.
+#' @family map variants
+#' @export
+#' @examples
+#' x <- list(list(a = 2:1, c = list(b1 = 2), b = list(c2 = 3, c1 = 4)))
+#' x |> str()
+#'
+#' # Transform each leaf
+#' x |> rmodify(f_leaf = \(x) x + 100) |>  str()
+#'
+#' # Recursively sort the nodes
+#' sort_named <- function(x) {
+#'   nms <- names(x)
+#'   if (!is.null(nms)) {
+#'     x[order(nms)]
+#'   } else {
+#'     x
+#'    }
+#' }
+#' x |> rmodify(f_post = sort_named) |> str()
+rmodify <- function(x,
+                    f_leaf = identity,
+                    p_leaf = NULL,
+                    f_pre = identity,
+                    f_post = identity) {
+  if (!is_vector(x)) {
+    cli::cli_abort("{.arg x} must be a vector, not {.obj_type_friendly {x}}.")
+  }
```

Comment:
I think it'd be more consistent if we checked for a leaf and return `leaf(x)` in that case.

Reply:
Why don't we require `x` to be a list with `vec_check_list()`?

I feel like allowing non-lists isn't useful because pre/post won't ever be applied? I guess that depends on what `p_leaf()` returns though (i.e. in theory it could say a numeric value is a node, but i feel like that isn't useful)

Reply:
I think there's probably some invariant around  `modify_tree(x, f_pre = \(y) modify_tree(y, f_leaf)) == modify_tree(x, f_leaf)`

Reply:
My preference would be to allow for as wide class of objects as possible. @lionel-'s point about allowing `rmodify` on all objects that `[[<-` is allowed on makes sense to me. 

Use cases: 
  - Side effectful recursive modification of nested environments
  - Recursive processing of language objects. For instance, if one wants to modify specific parts of a function.